Use return value from server to set prompt line

jgkamat opened the pull request:
This PR (finally!) adds support for reading the server return value to set the prompt line with the current instrument. 

Closes #1! :smile: 

I tried to add this in as cleanly as I could, but I could be overlooking something. Let me know if you see any problems or if you think this could be implemented better.

Also, I havent been able to reproduce the race condition I mentioned earlier with the new alda server! :tada:

jgkamat committed:
Add very basic query from alda server to get instruments

jgkamat committed:
Add ability to set prompt status from commands

jgkamat committed:
Simplify logic a bit

daveyarwood commented on src/alda/repl/AldaRepl.java:
```diff
@@ -144,18 +148,11 @@ public void run() {
+          if (playResponse != null && playResponse.currentInstrument() != null) {
+            // If we have multiple instruments, pick the first one.
+            String instrument = playResponse.currentInstrument();
+            this.setPromptPrefix(instrument);
```
In the Clojure Alda REPL we have now, we do something a little more complicated when displaying the current instruments:

- We set the prompt to the abbreviations of all of the current instruments, with slashes in between, e.g.: `p/t>` for a `piano` and a `trumpet`.

- If the instrument name contains dashes, we treat that as multiple "words" and the abbreviation is the first letter of each word, e.g. `midi-electric-piano-1` becomes `mep1`.

- If an instrument has a nickname, we use the nickname instead. For example, if you enter `trombone "foo": c`, the prompt reads `f>` for `foo`.

  These nicknames are stored as a map of nickname => list of instance IDs in the score map:

  ```clojure
  :nicknames
  {"foo" ("trumpet-KsU1s"),
   "strings" ("violin-A2i5P" "viola-YcVdU"),
   "strings.violin" ("violin-A2i5P"),
   "strings.viola" ("viola-YcVdU"),
   "baz" ("harmonica-4EOEB"),
   "bar" ("tuba-OttVs")},
   ```

It would be nice if  we can replicate this in the new REPL.

daveyarwood replied:
The Clojure implementation is here, for reference: https://github.com/alda-lang/alda-repl-clj/blob/master/src/alda/repl/core.clj#L60-L75

jgkamat replied:
oh I never knew that feature existed! I'll definetly add that in the next few commits! :smile:

daveyarwood commented on src/alda/repl/commands/ReplCommand.java:
```diff
@@ -13,16 +14,18 @@
    */
-  public void act(String args, StringBuffer history, AldaServer server, ConsoleReader reader);
+  public void act(String args, StringBuffer history, AldaServer server,
+                  ConsoleReader reader, Consumer<String> newInstrument);
```
I like your use of Consumer here, but I wonder if we could use it in a more general way? It seems like the intent is to update the prompt based on the results of the ReplCommand, so maybe it could be a `Consumer<AldaScore>` instead? That way, we could potentially adapt the Consumer to update the prompt in other ways, based on values in the score other than the current instruments.

jgkamat replied:
Yup, this is a good idea! (and it would reduce the shared json parsing across the individual scores).

daveyarwood reviewed (changes requested):
This is awesome, thanks for the PR! I made a few comments about things that could be tweaked.

I think we are really close to feature parity with the current REPL, which is great. Ideally, I'd like to be able to swap out the implementations in the next release of Alda, and it won't be obvious that it's a total rewrite, it will just magically start up faster! :smile:

daveyarwood commented:
Whoops -- I made a comment on an outdated diff -- it's still applicable, though. You may need to click "Show outdated" above to see it.

jgkamat committed:
Fix up behavior of instrument parsing to mirror clj alda repl

jgkamat commented:
The last commit should fix those issues above, but I wanted a way to test multiple instruments. Whats the easiest way to do that?

I _think_ the instrument parsing should be the same as the clj version now, let me know if you see any differences.

jgkamat committed:
Clean up default promptPrefix logic

daveyarwood commented:
I think you're on the right track, but I'm seeing one case that differs from the Clojure-based REPL -- 
`violin/cello "strings"` abbreviates as `s/s` (presumably both instruments are being abbreviated to `s` for `strings`), and in the Clojure REPL as `sc/sv>` (for `strings.cello` and `strings.violin`).

I think this is actually not working the way we want it in either version. Maybe the ideal behavior would be to just use the group nickname, i.e. `s>`.

I'm also occasionally seeing the prompt stop updating at all:

```
;; OK so far...
> violin/cello "strings": c
;; (although this should be s> as noted above)
s/s> bassoon "foo": c
;; f for foo
f> e
f> g
f> bassoon "bar": o2 c
;; b for bar
b> bassoon "quux": o2 g
;; q for quux
q> bar: a
[27713] ERROR Alda server is down. To start the server, run `alda up`.

;; At this point, we got an error because the response didn't come back within the
;; timeout allowed by the client code. I did still hear the "a" played by the bassoon
;; nicknamed "bar" shortly after the error printed.
;;
;; ...but now I'm in a weird state where the prompt stops updating:
q>
q>
q> bar: b
;; I hear a "b," but the prompt disappears!
>
>
> violin/viola/cello: c
;; I hear three "c's" in unison
>
>
> :new
> violin/viola/cello: c
;; I hear three "c's" in unison
> trombone: c
;; I hear a "c" on trombone
> flute: c c c c c c c
;; I hear a bunch of "c's" on flute
>
```

I think there are some issues I need to address from the server side. I'll make some GitHub issues and work on them in the near future whenever I can find pockets of time.

I have a request: do you think you could add a `:debug` REPL command that will enable a mode where we print all of the AldaResponse JSON strings as they are received? Maybe it could take an argument like `:debug on` or `:debug off` to enable/disable it. This would help me a lot as I work on fixing bugs on the server side.

jgkamat commented:
I'll add the `:debug` command asap (since it definitely would help debugging things server side). I'm not sure why the prompt stopped updating, the default when there's some error is to clear the prompt, but it really should come back when a successful run happens, I probably need to take a closer look at that. I might not be able to fix this for a while (the rest of the behavior, but I will get to it eventually (probably next weekend) :smile:)

jgkamat committed:
Add :debug command to toggle debug status

Very quick patch, needs cleanup later.
Eventually should take a on/off (or true/false) flag.

daveyarwood commented:
Awesome, thanks for hacking that in!

jgkamat committed:
Merge branch 'master' of github.com:alda-lang/alda-client-java into jay/repl-status

jgkamat commented:
@daveyarwood I think I may have figured out why we're seeing that issue (and while I don't have a good solution right now, I think you should know about it so we don't waste time trying to debug it). 

This problem is actually on the client side: when we have a failed request (example: server not there to receive), something in the client (I'm thinking something in ZMQ) keeps trying to contact the server (for an indefinite amount of time). When the server does come back up, that request finally goes through (the one that 'failed' when the server went down), and the server sends back a 'not enough workers' reply. This goes in some sort of queue that is later pulled from when we actually read. From that point in the future, all of our responses lag behind by one request (which is what tripps up the prompt, and anything else relying on responses). 

A very dirty way to fix this is this:

```
diff --git a/src/alda/AldaRequest.java b/src/alda/AldaRequest.java
index 17cfebf4aa57e4558c77d68eb4f91adb2b4a2bd4..d714472ded6c1eaad01e58fc350017a5738b36c9 100644
--- a/src/alda/AldaRequest.java
+++ b/src/alda/AldaRequest.java
@@ -19,7 +19,8 @@ public class AldaRequest {
   }
 
   private Socket getDealerSocket() {
-    Socket socket = findExistingSocketForHostAndPort(this.host, this.port);
+    // Socket socket = findExistingSocketForHostAndPort(this.host, this.port);
+    Socket socket = null;
     if (socket != null){
       return socket;
     }
```

I'm going to keep looking at this to see if I can find a cleaner solution than that though. I think the best simple solution is to trash the existing socket when we ever have a connection problem (which should prevent us from ever getting queued responses). If you know any setting in ZMQ that forces these requests to timeout let me know though... 

This is probably not worth the effort, but we could have a system where responses are handled independently in their own thread (in the client)...

daveyarwood commented:
> I think the best simple solution is to trash the existing socket when we ever have a connection problem (which should prevent us from ever getting queued responses). 

I think this is actually the approach recommended by the ZeroMQ Guide: see their [lpclient.java](http://zguide.zeromq.org/java:lpclient) example. I wasn't really sure about that approach because when I first tried it, I ended up creating too many file descriptors and getting a "Too many open files" error. Since then, though, I've learned more about how to use ZeroMQ correctly, and there have also been some bug fixes in the library we use, so I feel pretty good about trying this approach again. I'll give it a try soon.

daveyarwood commented:
There is also this issue on the server side: https://github.com/alda-lang/alda-server-clj/issues/8

The approach I'd like to take is for the server to take a "job ID" from the client on each request and maintain a cache of previous jobs, each one containing the status of that request, the score map, and an error message if there was an error. That way, we wouldn't have to rely on the order of messages coming off the queue. The client would simply retry its request until the server sends back a response for the correct job.

I do think destroying the socket on each retry is also a good idea, though.

jgkamat committed:
Clean up sockets when a connection error occurs

jgkamat committed:
Tweak prefix logic to prefer a single nickname if possible

jgkamat committed:
Clean up and improve :debug command

daveyarwood committed:
update jeromq => 0.4.0; minor improvements to AldaRequest

- use a Poller created via the context, so that polling resources are reaped
  when the context is closed

- reuse the same poller, so we aren't creating a new selector on every retry

- use the singleton pattern to eliminate the context and socket arguments

- rebuild the socket on each retry

daveyarwood committed:
include a randomly generated UUID on each play/play-status request

I have a patch coming through in alda-server-clj that separates jobs by ID,
and ensures that the response a worker sends corresponds to the request it
received for that score. The worker sends back the same jobId in its
response, so the client can read that and ignore any responses that aren't for
the request it sent.

jgkamat committed:
Merge pull request #1 from alda-lang/dave/repl-status

A couple updates

daveyarwood commented:
I think this branch is in good shape -- excellent work @jgkamat :beers:

I'm going to go ahead and merge it -- there is still a little bit of weirdness around the server not necessarily responding right away, leading to timeouts, which I've captured in https://github.com/alda-lang/alda-client-java/issues/12. I'll take a look at improving things there.
